File: src/cli/agent/truncation.ts
/**
 * Context-safe truncation for agent output.
 *
 * Identical semantics to the original truncation module but with NO mutable
 * globals.  Full-output dumps use node:fs directly — the truncation path is
 * a best-effort side-channel that doesn't need to go through Effect layers.
 */

import * as fs from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

const MAX_LIST_ITEMS = 50;
const MAX_STRING_LENGTH = 1000;
const MAX_SERIALIZED_BYTES = 16 * 1024;

export interface TruncationMetadata {
  truncated: boolean;
  total: number;
  shown: number;
  full_output?: string;
}

export interface ListTruncationResult<T> {
  items: T[];
  metadata: TruncationMetadata;
}

export interface PayloadTruncationResult<T> {
  value: T;
  metadata?: TruncationMetadata;
}

function estimateBytes(value: unknown): number {
  return Buffer.byteLength(JSON.stringify(value), "utf8");
}

function slugify(commandId: string): string {
  return commandId.replace(/[^a-zA-Z0-9-_.]+/g, "-");
}

export function writeFullOutput(commandId: string, payload: unknown): string {
  const dir = join(tmpdir(), "godaddy-cli");
  fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
  try {
    fs.chmodSync(dir, 0o700);
  } catch {
    // Best-effort on platforms that don't honor POSIX modes.
  }
  const filename = `${Date.now()}-${slugify(commandId)}.json`;
  const fullPath = join(dir, filename);
  fs.writeFileSync(fullPath, JSON.stringify(payload, null, 2), {
    encoding: "utf8",
    mode: 0o600,
  });
  try {
    fs.chmodSync(fullPath, 0o600);
  } catch {
    // Best-effort on platforms that don't honor POSIX modes.
  }
  return fullPath;
}

function truncateStrings(value: unknown): unknown {
  if (typeof value === "string") {
    if (value.length <= MAX_STRING_LENGTH) {
      return value;
    }
    return `${value.slice(0, MAX_STRING_LENGTH)}...(truncated)`;
  }

  if (Array.isArray(value)) {
    return value.map((item) => truncateStrings(item));
  }

  if (value && typeof value === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, nested] of Object.entries(value)) {
      result[key] = truncateStrings(nested);
    }
    return result;
  }

  return value;
}

export function truncateList<T>(
  items: T[],
  commandId: string,
): ListTruncationResult<T> {
  const total = items.length;
  const shown = Math.min(total, MAX_LIST_ITEMS);
  const truncated = total > MAX_LIST_ITEMS;
  const sliced = truncated ? items.slice(0, MAX_LIST_ITEMS) : items;

  if (!truncated) {
    return {
      items: sliced,
      metadata: { truncated: false, total, shown },
    };
  }

  const fullOutput = writeFullOutput(commandId, items);
  return {
    items: sliced,
    metadata: { truncated: true, total, shown, full_output: fullOutput },
  };
}

export function protectPayload<T>(
  value: T,
  commandId: string,
): PayloadTruncationResult<T> {
  const totalBytes = estimateBytes(value);
  let candidate = truncateStrings(value) as T;
  let shownBytes = estimateBytes(candidate);
  let truncated = totalBytes !== shownBytes;
  let fullOutput: string | undefined;

  if (shownBytes > MAX_SERIALIZED_BYTES) {
    truncated = true;
    const limited = {
      truncated: true,
      summary:
        typeof value === "object" && value !== null
          ? "Output too large for inline payload"
          : String(value),
    };
    candidate = limited as T;
    shownBytes = estimateBytes(candidate);
  }

  if (truncated) {
    fullOutput = writeFullOutput(commandId, value);
    return {
      value: candidate,
      metadata: {
        truncated: true,
        total: totalBytes,
        shown: shownBytes,
        full_output: fullOutput,
      },
    };
  }

  return { value: candidate };
}

File: rust/src/truncation.rs
//! Context-safe truncation for agent-facing command output.
//!
//! Large lists and payloads are capped before being rendered so a single
//! command can't blow an agent's context window; the untruncated data is
//! written to a side-channel file under `$TMPDIR/godaddy-cli/` instead.

use std::io::Write;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

use serde_json::Value;

const MAX_STRING_LENGTH: usize = 1000;
const TRUNCATION_SUFFIX: &str = "...(truncated)";
const MAX_SERIALIZED_BYTES: usize = 16 * 1024;

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct TruncationMetadata {
    pub(crate) truncated: bool,
    pub(crate) total_bytes: usize,
    pub(crate) shown_bytes: usize,
    pub(crate) full_output: Option<String>,
}

pub(crate) struct PayloadTruncationResult {
    pub(crate) value: Value,
    pub(crate) metadata: Option<TruncationMetadata>,
}

fn slugify(command_id: &str) -> String {
    command_id
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '.') {
                c
            } else {
                '-'
            }
        })
        .collect()
}

/// Counts bytes written without buffering them, so measuring a large value's
/// serialized size doesn't require allocating a same-sized `String` first.
struct CountingWriter(usize);

impl Write for CountingWriter {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        self.0 += buf.len();
        Ok(buf.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        Ok(())
    }
}

fn estimate_bytes(value: &Value) -> usize {
    let mut writer = CountingWriter(0);
    serde_json::to_writer(&mut writer, value).ok();
    writer.0
}

/// Best-effort dump of the untruncated payload to a `0600` file under a
/// `0700` directory. Returns `None` if the filesystem write fails, in which
/// case the caller simply omits `full_output` from the truncation metadata.
fn write_full_output(command_id: &str, payload: &Value) -> Option<PathBuf> {
    let dir = std::env::temp_dir().join("godaddy-cli");
    std::fs::create_dir_all(&dir).ok()?;
    #[cfg(unix)]
    {
        use std::fs::Permissions;
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&dir, Permissions::from_mode(0o700)).ok();
    }

    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .ok()?
        .as_millis();
    let path = dir.join(format!(
        "{millis}-{}-{}.json",
        slugify(command_id),
        uuid::Uuid::new_v4()
    ));
    let contents = serde_json::to_string_pretty(payload).ok()?;
    std::fs::write(&path, contents).ok()?;
    #[cfg(unix)]
    {
        use std::fs::Permissions;
        use std::os::unix::fs::PermissionsExt;
        std::fs::set_permissions(&path, Permissions::from_mode(0o600)).ok();
    }

    Some(path)
}

fn truncate_strings(value: &Value) -> Value {
    match value {
        Value::String(s) if s.chars().count() > MAX_STRING_LENGTH => {
            let take_len = MAX_STRING_LENGTH.saturating_sub(TRUNCATION_SUFFIX.chars().count());
            let truncated: String = s.chars().take(take_len).collect();
            Value::String(format!("{truncated}{TRUNCATION_SUFFIX}"))
        }
        Value::Array(items) => Value::Array(items.iter().map(truncate_strings).collect()),
        Value::Object(obj) => Value::Object(
            obj.iter()
                .map(|(key, nested)| (key.clone(), truncate_strings(nested)))
                .collect(),
        ),
        other => other.clone(),
    }
}

/// Truncates long strings within `value`, then falls back to a `{ truncated,
/// summary }` stub if the result still exceeds [`MAX_SERIALIZED_BYTES`].
/// Dumps the original, untouched `value` to a side-channel file whenever
/// truncation occurred.
pub(crate) fn protect_payload(value: Value, command_id: &str) -> PayloadTruncationResult {
    let total_bytes = estimate_bytes(&value);
    let mut candidate = truncate_strings(&value);
    let mut shown_bytes = estimate_bytes(&candidate);
    let mut truncated = total_bytes != shown_bytes;

    if shown_bytes > MAX_SERIALIZED_BYTES {
        truncated = true;
        candidate = serde_json::json!({
            "truncated": true,
            "summary": "Output too large for inline payload",
        });
        shown_bytes = estimate_bytes(&candidate);
    }

    if !truncated {
        return PayloadTruncationResult {
            value: candidate,
            metadata: None,
        };
    }

    let full_output = write_full_output(command_id, &value).map(|path| path.display().to_string());
    PayloadTruncationResult {
        value: candidate,
        metadata: Some(TruncationMetadata {
            truncated: true,
            total_bytes,
            shown_bytes,
            full_output,
        }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn truncate_strings_truncates_long_values_recursively() {
        let long = "a".repeat(1500);
        let value = json!({ "nested": [long.clone()] });
        let truncated = truncate_strings(&value);
        let nested = truncated["nested"][0].as_str().expect("string");
        assert!(nested.ends_with(TRUNCATION_SUFFIX));
        assert_eq!(nested.chars().count(), MAX_STRING_LENGTH);
    }

    #[test]
    fn protect_payload_passes_through_small_values_unchanged() {
        let value = json!({ "name": "short" });
        let result = protect_payload(value.clone(), "test-small");
        assert_eq!(result.value, value);
        assert!(result.metadata.is_none());
    }

    #[test]
    fn protect_payload_truncates_long_strings_and_reports_metadata() {
        let long = "a".repeat(1500);
        let value = json!({ "name": long });
        let result = protect_payload(value, "test-strings");
        let metadata = result.metadata.expect("metadata present");
        assert!(metadata.truncated);
        let full_output = metadata.full_output.expect("full_output path");
        std::fs::remove_file(full_output).ok();
    }

    #[test]
    fn protect_payload_falls_back_to_summary_when_still_too_large() {
        let big_array: Vec<String> = (0..2000).map(|i| format!("item-{i}")).collect();
        let value = json!({ "items": big_array });
        let result = protect_payload(value, "test-huge");
        let metadata = result.metadata.expect("metadata present");
        assert!(metadata.truncated);
        assert_eq!(result.value["truncated"], json!(true));
        assert!(result.value.get("summary").is_some());
        let full_output = metadata.full_output.expect("full_output path");
        std::fs::remove_file(full_output).ok();
    }
}
